src/libraries/System.Private.CoreLib/src/System/Diagnostics/DebugProvider.cs

```diff
@@ -25,7 +28,9 @@ public virtual void Fail(string? message, string? detailMessage)
             }
             WriteAssert(stackTrace, message, detailMessage);
             FailCore(stackTrace, message, detailMessage, "Assertion failed.");
+#nullable disable
```

Comment:
:memo: without this suppression, the compiler points out that this method could return normally. Looking at the implementation of `FailCore`, I think it has the same problem.
We need to discuss what is the best way to suppress this warning. Maybe a more fined-grained suppression on a specific diagnostic would be better than the coarse `#nullable disable` I used...

Reply:
You're nullable disabling/enabling around the closing brace?

Regardless, yes, it's possible for someone to implement Fail in a way that still returns, but that goes against the purpose of the method.  If we held true to the possibility that someone did that, then we'd need to remove `[DoesNotReturnIf(false)]` from Debug.Assert and all such methods, which would be terrible.

Reply:
Can't we add `DoesNotReturn` to `FailCore`?

Reply:
@safern  `FailCore` has `return` statements, so the same warning "method should not return" would be produced there if we add `[DoesNotReturn]`.